app/src/main/java/org/mozilla/tv/firefox/session/SessionObservationManager.kt

```diff
@@ -0,0 +1,56 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.session
+
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+
+/**
+ * TODO
+ * when sessions are added or removed, add/remove session observers
+ * when sessions change, update repo
+ */
+class SessionObservationManager private constructor(sessionRepo: SessionRepo) { // TODO rename this
+
+    companion object {
+        fun attach(sessionRepo: SessionRepo, sessionManager: SessionManager) {
+            val updater = SessionObservationManager(sessionRepo)
+            sessionManager.selectedSession?.register(updater.sessionObserver)
+            sessionManager.register(updater.sessionManagerObserver)
+        }
+    }
+
+    // Updates the repo whenever the observed session changes
+    private val sessionObserver = object : Session.Observer {
+        override fun onUrlChanged(session: Session, url: String) {
+            sessionRepo.update()
+        }
+
+        override fun onDesktopModeChanged(session: Session, enabled: Boolean) {
+            sessionRepo.update()
+        }
+
+        override fun onLoadingStateChanged(session: Session, loading: Boolean) {
+            sessionRepo.update()
+        }
+
+        override fun onNavigationStateChanged(session: Session, canGoBack: Boolean, canGoForward: Boolean) {
+            sessionRepo.update()
+        }
+    }
+
+    // Attaches and removes the session observer whenever sessions are added
+    // and removed
+    val sessionManagerObserver = object : SessionManager.Observer {
+        override fun onSessionSelected(session: Session) {
+            session.register(sessionObserver)
```

Comment:
It's unclear to me exactly whether a session can a) not be the selectedSession but b) not removed - will we be observing multiple sessions? I guess that's fine bc they don't receive updates (and for us right now, we only have the one session).

Reply:
My understanding is that we will always have 0 or 1 sessions, but never more.  If that changes in the future I assume a lot of weird edge cases will be dredged up, but you're probably right that it's something we should think about sooner rather than later.